File: modules/@angular/platform-browser/test/web_workers/shared/message_bus_util.dart
library angular2.test.web_workers.shared.message_bus_util;

import 'dart:isolate';
import 'package:angular2/src/web_workers/shared/message_bus.dart'
    show MessageBus;
import 'package:angular2/src/web_workers/shared/isolate_message_bus.dart';

/*
 * Returns an IsolateMessageBus thats sink is connected to its own source.
 * Useful for testing the sink and source.
 */
MessageBus createConnectedMessageBus() {
  var receivePort = new ReceivePort();
  var sendPort = receivePort.sendPort;

  var sink = new IsolateMessageBusSink(sendPort);
  var source = new IsolateMessageBusSource(receivePort);

  return new IsolateMessageBus(sink, source);
}

File: modules/@angular/platform-webworker/test/web_workers/shared/message_bus_util.ts
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */

import {MessageBus} from '@angular/platform-webworker/src/web_workers/shared/message_bus';
import {PostMessageBus, PostMessageBusSink, PostMessageBusSource} from '@angular/platform-webworker/src/web_workers/shared/post_message_bus';


/*
 * Returns a PostMessageBus thats sink is connected to its own source.
 * Useful for testing the sink and source.
 */
export function createConnectedMessageBus(): MessageBus {
  const mockPostMessage = new MockPostMessage();
  const source = new PostMessageBusSource(<any>mockPostMessage);
  const sink = new PostMessageBusSink(mockPostMessage);

  return new PostMessageBus(sink, source);
}

class MockPostMessage {
  private _listener: EventListener;

  addEventListener(type: string, listener: EventListener, useCapture?: boolean): void {
    if (type === 'message') {
      this._listener = listener;
    }
  }

  postMessage(data: any, transfer?: [ArrayBuffer]): void { this._listener(<any>{data: data}); }
}
